tap_dynamodb/deserialize.py

```diff
@@ -58,23 +58,61 @@ def _apply_projection(self, record, breadcrumb, output):
+                        output[breadcrumb_key].append(record[breadcrumb_key][index])
                 else:
-                    output[breadcrumb_key] = [record[breadcrumb_key][index]]
+                    # record: {'Artist': 'No One You Know5'}
+                    # main breadcrumb = [['metadata[0]']]
+                    # current breadcrumb = ['metadata[0]']
+                    # current output = {}
+                    # expected output = {'metadata': []}
+                    #       as "metadata" does not have any items and the current breadcrumb is expecting 1st item
 
+                    output[breadcrumb_key] = []
+                    # only prepare output if the list field contains data at that index position in record
+                    if record.get(breadcrumb_key) and len(record.get(breadcrumb_key)) > index:
+                        output[breadcrumb_key].append(record[breadcrumb_key][index])
             else:
                 output[breadcrumb[0]] = record.get(breadcrumb[0])
         else:
             if '[' in breadcrumb[0]:
                 breadcrumb_key = breadcrumb[0].split('[')[0]
                 index = int(breadcrumb[0].split('[')[1].split(']')[0])
-                if output.get(breadcrumb_key) is None:
+                if not output.get(breadcrumb_key):
                     output[breadcrumb_key] = [{}]
-                self._apply_projection(record[breadcrumb_key][index], breadcrumb[1:], output[breadcrumb_key][0])
+
+                # record: {'Artist': 'No One You Know5'}
+                # main breadcrumb = [['metadata[0]', 'Age']]
+                # current breadcrumb = ['metadata[0]', 'Age']
+                # current output = {'metadata': [{}]}
+                # expected output = {'metadata': [{}]}
+                #       as "metadata" is not present and the current breadcrumb is expecting 1st item and which is a parent
+
+                # only prepare output if the list field contains data at that index position in record
+                if record.get(breadcrumb_key) and len(record.get(breadcrumb_key)) > index:
+                    self._apply_projection(record[breadcrumb_key][index], breadcrumb[1:], output[breadcrumb_key][0])
             else:
                 if output.get(breadcrumb[0]) is None:
                     output[breadcrumb[0]] = {}
-                self._apply_projection(record[breadcrumb[0]], breadcrumb[1:], output[breadcrumb[0]])
+
+                # record: {'Artist': 'No One You Know5'}
+                # main breadcrumb = [['metadata', 'inner_metadata']]
+                # current breadcrumb = ['metadata', 'inner_metadata']
+                # current output = {'metadata': {}}
+                # expected output = {'metadata': {}}
+                #       as "metadata" is not present and the current breadcrumb is expecting it as a parent
+
+                # keep empty dict if the data is not found in the record
+                if record.get(breadcrumb[0]):
```

Comment:
```suggestion
                if breadcrumb[0] in record:
```

Reply:
Updated.

Reply:
@cosimon
![image](https://user-images.githubusercontent.com/75604662/149314967-47ef3e8f-bcbf-496d-8245-a3b3f745549f.png)

As seen in the figure, the particular code was kept to handle such scenarios where the dictionary contains an empty list which would return false if we change it to if breadcrumb_key not in output:
Hence, not updated the code.